sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -324,14 +324,22 @@ object TypeCoercion {
       // We should cast all relative timestamp/date/string comparison into string comparisons
       // This behaves as a user would expect because timestamp strings sort lexicographically.
       // i.e. TimeStamp(2013-01-01 00:00 ...) < "2014" = true
-      case p @ BinaryComparison(left @ StringType(), right @ DateType()) =>
-        p.makeCopy(Array(left, Cast(right, StringType)))
-      case p @ BinaryComparison(left @ DateType(), right @ StringType()) =>
-        p.makeCopy(Array(Cast(left, StringType), right))
-      case p @ BinaryComparison(left @ StringType(), right @ TimestampType()) =>
-        p.makeCopy(Array(left, Cast(right, StringType)))
-      case p @ BinaryComparison(left @ TimestampType(), right @ StringType()) =>
-        p.makeCopy(Array(Cast(left, StringType), right))
+      // If StringType is foldable then we need to cast String to Date or Timestamp type
+      // which would give order of magnitude performance gain as well as preserve the behavior
+      // achieved by expressed above
+      // TimeStamp(2013-01-01 00:00 ...) < Cast( "2014" as timestamp) = true
+      case p @ BinaryComparison(left @ StringType(), right) if dateOrTimestampType(right) =>
+        if (left.foldable) {
+          p.makeCopy(Array(Cast(left, right.dataType), right))
```

Comment:
You are changing the semantics of the comparison here. For example `'10' < current_timestamp` currently returns `true`, if we apply your PR it will return `null`. This is a breaking change, and I don't think we should do this.